### ISSUE TITLE
RejectedExecutionException for AmqpsDeviceAuthenticationCBSTokenRenewalTask

### ISSUE THREAD

**Issue by rkanubaddhi** [Azure/azure-iot-sdk-java#294]
- **OS and version used:** Windows 7 

- **Java runtime used:** 1.8 

- **SDK version used:** 1.11.0 

# Description of the issue:
I'm seeing following exception once in a while, not sure why AmqpsDeviceAuthenticationCBSTokenRenewalTask is getting rejected. I don't see any issues even though I'm seeing this exception. I still have bidirectional communication. As it is a warning level logging, I want to make sure everything still works fine after this warning message.

I would like to understand in what scenarios I might see the following exception?

WARNING [10:08:28 02-Jul-18 CDT][com.microsoft.azure.sdk.iot.device.transport.amqps.AmqpsDeviceAuthenticationCBSTokenRenewalTask] java.util.concurrent.RejectedExecutionException: Task java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask@17d990 rejected from java.util.concurrent.ScheduledThreadPoolExecutor@d03b62[Shutting down, pool size = 1, active threads = 1, queued tasks = 0, completed tasks = 1]: Task java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask@17d990 rejected from java.util.concurrent.ScheduledThreadPoolExecutor@d03b62[Shutting down, pool size = 1, active threads = 1, queued tasks = 0, completed tasks = 1]

# Need Support?
* Have a feature request for SDKs? Please post it on [User Voice](https://feedback.azure.com/forums/321918-azure-iot) to help us prioritize.
* Have a technical question? Ask on [Stack Overflow](https://stackoverflow.com/questions/tagged/azure-iot-hub) with tag “azure-iot-hub”
* Need Support? Every customer with an active Azure subscription has access to support with guaranteed response time.  Consider submitting a ticket and get assistance from Microsoft support team
* Found a bug? Please help us fix it by thoroughly documenting it and filing an issue on GitHub (C, Java, .NET, Node.js, Python).

**Comment by timtay-microsoft** [comment]
The thread that is being rejected is used to refresh your authorization credentials over the AMQP connection periodically so that you never have to do so manually. Are you experiencing any connectivity issues anytime after this warning appears?

**Comment by rkanubaddhi** [comment]
No, I don't see any connectivity issues after the warning. I see that thread getting rejected even when I have a valid SaS token, I'm not sure how to reproduce it, it is not happening consistently. In order to reproduce I wanted to understand the scenarios when the thread could be rejected.

**Comment by rkanubaddhi** [comment]
Also is there anyway to disable thread - AmqpsDeviceAuthenticationCBSTokenRenewalTask. We have our own way of renewing the SaS token.

**Comment by timtay-microsoft** [comment]
Do you mean that when you create your device client, you are only providing a sas token, not the device key? Currently, we highly recommend you provide the whole device key to allow the cbs task to handle token renewal. 

If you still want to handle the token renewal yourself, your amqp connection will be lost when your token expires, and you will have to open a new connection yourself by creating an entire other Device Client. In this case, the warning log you are seeing doesn't matter at all, since the cbs task is useless since it will never be able to refresh your sas token for you.

Currently, we don't expose a way to disable this cbs task even though it is redundant in your case.

**Comment by rkanubaddhi** [comment]
Is there anyway to just disable the warning message? It is confusing to see the warning message even though we have a full connection.

**Comment by timtay-microsoft** [comment]
All you can do is change the log level of your logger to exclude warning level messages.

**Comment by rkanubaddhi** [comment]
That will not log all other warning level messages, is there anyway to disable only the above message?

**Comment by rkanubaddhi** [comment]
Thanks for the quick response, back to the original question when this thread will be rejected? why is it being rejected? It is okay to execute AmqpsDeviceAuthenticationCBSTokenRenewalTask, when though we have our own renewal task.

**Comment by timtay-microsoft** [comment]
I'm not sure why exactly the thread is being rejected. I'd have to spend some time to dig into it. It is perfectly safe to allow this thread to run even when you renew the sas token manually. Eventually, this task will throw an exception to alert that the existing sas token has expired, and your amqp connection will close as a result.

**Comment by rkanubaddhi** [comment]
In our implementation we are renewing the sas before it expires, so sas token getting expired is not a problem, once you dig into it, please let me know in what scenarios thread gets rejected. Thanks for the quick response!

**Comment by rkanubaddhi** [comment]
any update on this?

**Comment by timtay-microsoft** [comment]
We haven't had much time to look into this lately, sadly. I'll update this thread once we make some progress.

**Comment by timtay-microsoft** [comment]
Sorry for the delay on this issue. I was able to reproduce this behavior on my side using Device Client 1.11.0. The fix has been checked in as of 9fd5db1499f82db07cce609fe41e604fa4e8b998
